File: 15.three_sum.rb
def three_sum nums
  @result = []
  @nums = nums.sort!
  find_result
  @result.uniq
end

def find_result
  @nums.each_with_index do |target, index|
    two_sum target, index
  end
end

def two_sum target, index
  element = []
  (index + 1..@nums.size - 1).each do |i|
    element.include?(-target - @nums[i]) ? @result << [target, -target - @nums[i], @nums[i]] : element << @nums[i]
  end
end

print three_sum([-1, 0, 2, 3, -5, 1, -4, 0, 0, 0, 1, 2]), "\n"

File: 15.ThreeSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

// 311/48ms/11.70%
class ThreeSum {
  public List<List<Integer>> threeSum(int[] nums) {
    Arrays.sort(nums);
    List<List<Integer>> list = new ArrayList<>();
    Set<Integer> targets = new HashSet<>();

    for (int target = 0; target < nums.length; target++) {
      if (!targets.contains(nums[target])) {
        Set<Integer> set = new HashSet<>();
        List<Integer> previousList = new ArrayList<>();

        for (int i = target + 1; i < nums.length; i++) {
          if (set.contains(-nums[target] - nums[i])) {
            List<Integer> innerList = new ArrayList<>();
            innerList.add(nums[target]);
            innerList.add(-nums[target] - nums[i]);
            innerList.add(nums[i]);

            if (!innerList.equals(previousList)) {
              list.add(innerList);
              previousList = innerList;
            }
          } else {
            set.add(nums[i]);
          }
        }
        targets.add(nums[target]);
      }
    }
    return list;
  }

  public static void main(String[] args) {
    System.out.println(new ThreeSum().threeSum(new int[] {4, 0, -1, -1, -1, 2, 2, 2, 2, 0, 0, 0, -4}));
  }
}